string.c

```diff
@@ -6161,9 +6161,8 @@ str_byte_substr(VALUE str, long beg, long len, int empty)
     else
         p = s + beg;
 
-    if (!STR_EMBEDDABLE_P(len, TERM_LEN(str)) && SHARABLE_SUBSTRING_P(beg, len, n)) {
-        str2 = rb_str_new_frozen(str);
-        str2 = str_new_shared(rb_cString, str2);
+    if (!STR_EMBEDDABLE_P(len, TERM_LEN(str)) && SHARABLE_SUBSTRING_P(beg, len, n) && OBJ_FROZEN(str)) {
```

Comment:
Trying to get the reasoning here, based on the limited context on the issue.

The assumption here is that if `str` (aka original string) is mutable, if we make it a shared string, we'll need to copy the whole string again if we mutate it. Is that right?

Reply:
Yeah, I think that's the logic.

We basically end up duplicating and creating a frozen shared string, only to essentially throw the original string away. Better to freeze the original before splitting it.

Reply:
IIRC:

```ruby
# Accumulate data from the network:
buffer = io.read # read some data from the Network

# Option 1: String.byteslice creates an internal shared copy which is never really reused:
message = string.byteslice(0, 25)
buffer = buffer.byteslice(25, buffer.size)

# Option 2: Freeze the buffer before splitting it.
buffer.freeze # Allow this buffer to be shared.
message = string.byteslice(0, 25)
buffer = buffer.byteslice(25, buffer.size)
```

Reply:
Right, so it's somewhat a similar problem to https://github.com/ruby/ruby/pull/5593

The hard part with these kind of changes, is that these optimizations are based on assumptions that are basically impossible to say whether they old true. Like I totally agree that this behavior isn't desirable when you do the classic buffer slicing code. But maybe this would degrade performance in other patterns? That's very hard to say.

I can't really think of patterns that would benefit of this though, so I'd say 👍 , but might be worth involving a couple more people to see if they can think of situations where this could degrade performance.

Reply:
Also there are probably other condition that may apply here, e.g.:

  - if `str` is a shared string, we can share again safely.
  - if `str` is hidden but somehow not frozen (is that even possible).

Reply:
As discussed on slack, we actually found a pattern for which this would degrade the performance:

  - https://github.com/redis/redis-rb/blob/4e9d73d3bb47831fe720cbce7c47cb11dd3f4de9/lib/redis/connection/ruby.rb#L96
  - https://github.com/redis-rb/redis-client/blob/2d51d6a077f87181dff500fcff6fee7dfece0ad2/lib/redis_client/ruby_connection/buffered_io.rb#L68

But that code could preserve the performance by freezing the buffer.